bigflow/konfig.py

```diff
@@ -0,0 +1,280 @@
+    __slots__ = ('__dict__', '_frozen', '_init_kwargs')
+
+    def __init__(self, **kwargs):
+        self._init_kwargs = kwargs
+
+    def __post_init__(self, **kwargs):
+        for k in dir(self):
+            if not k.startswith("_"):
+                # "materialize" all 'cached_property's
+                getattr(self, k)
+
+    def __setattr__(self, name, value):
+        if getattr(self, '_frozen', False) and name != '_frozen':
+            raise RuntimeError("Attempt to modify frozen Konfig")
+        object.__setattr__(self, name, value)
+
+    def __repr__(self):
+        cls = type(self)
+        kvpairs_list = ", ".join(f"{k}={v!r}" for k, v in self.items())
+        return f"{cls.__module__}.{cls.__qualname__}({kvpairs_list})"
+
+    @classmethod
+    def _make(cls: tp.Type[K], kvpairs: tp.List[tp.Tuple[str, tp.Any]]) -> K:
+        return cls(**dict(kvpairs))
+
+    def __reduce__(self):
+        return (
+            self._make,
+            (list(self.__dict__.items()),),
+        )
+
+    # adapt to `collections.abc.Mapping`
+
+    def __getitem__(self, k):
+        return getattr(self, k)
+
+    def __iter__(self):
+        return iter(vars(self).keys())
+
+    def __len__(self):
+        return len(vars(self))
+
+    # helpers & utils
+
+    @property
+    def name(self):
+        return type(self)._name
+
+    def replace(self, **kwargs):
+        """Return a new instance of the konfig replacing specified fields"""
+        cls = type(self)
+        return cls(**{**self._init_kwargs, **kwargs})
+
+    @classmethod
+    def resolve(
+        cls: tp.Type[K],
+        name: tp.Optional[str] = None,
+        default: tp.Optional[str] = None,
+        lazy: bool = True,
+        extra: tp.Optional[tp.Dict[str, tp.Any]] = None,
+    ) -> K:
+        return resolve_konfig(
+            [k for k in _all_subclasses(cls) if k._name],
+            name=name,
+            default=default,
+            lazy=lazy,
+            extra=extra,
+        )
+
+
+def _all_subclasses(cls: type):
+    yield cls
+    for sc in cls.__subclasses__():
+        yield from _all_subclasses(sc)
+
+
+class _LazyKonfig(lazy_object_proxy.Proxy):
+
+    def __init__(self, **kwargs):
+        super().__init__(lambda: resolve_konfig(**kwargs))
+
+    def __repr__(self):
+        return str(self)
+
+
+def resolve_konfig(
+    konfigs: tp.Union[
+        tp.Dict[str, tp.Type[K]],
+        tp.List[tp.Type[K]],
+    ],
+    name: tp.Optional[str] = None,
+    default: tp.Optional[str] = None,
+    lazy: bool = True,
+    extra: tp.Optional[tp.Dict[str, tp.Any]] = None,
+) -> K:
+
+    """Creates instance of config.
+
+    Konfigs may be passed as list (in such case Class.name is used as a konfig name)
+    or as a plain dictionary {name -> konfig_class}.
+
+    Konfig name may be provided explicitly with `name` argument or passed via `bigflow` cli tool.
+    Default konfig config name may be passed as a fallback.
+    """
+
+    if not isinstance(konfigs, tp.Dict):
```

Comment:
Im not sure if having 2 mechanisms for the same task is a good idea. I would pick the one with the argument in the class definition.

Reply:
I'd prefer to keep only 'dict' variant. It is a bit more flexible (you can always redefine konfigs), less magical,
and also clearly shows what environments do you have (no need to search through the code). WDYT?

Reply:
Agree, keep the dict based mechanism

Reply:
done